Wait for processes to be reaped if you are doing `rm -f`

ibuildthecloud opened the pull request:
If you do `rm -f` at just the right time when a container PID 1 has died but has not yet been reaped by Docker/Linux then kill will return a ESRCH.  In this situation we wait a bit for Docker/Linux to cleanup so we can delete the container.

As this is a race condition it is very difficult to reproduce but can be done by writing a program that uses the Docker API and does the equilavent of `docker rm -f $(docker run -d busybox echo hi)`.  Running that on the command line will never fail, but a program is fast enough that it will hit the situation a good amount of times.

One thing I'm not too sure about this fix is whether a new error should be introduced.  I added a an error `execdriver.ErrProcessNotFound` to catch this situation.  But this now means that `driver.Kill()` doesn't return a `syscall.Errno` for this scenario.  Would it be preferable to not mask the error?  I couldn't make the judgment call myself.  The question is whether you want the caller of Driver.Kill() to expect syscall.Errno or define new errors in the `execdriver` package.

LK4D4 commented:
@ibuildthecloud I vote for not masking :)

LK4D4 commented on daemon/delete.go:
```diff
@@ -55,7 +57,19 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
+						// This is a situation in which the container PID 1 has died
+						// but has yet to be reaped or recognized by Docker as being gone.
+						// wait for the container to stop
+						if _, err := container.WaitStop(10 * time.Second); err != nil {
```
Is it really necessary to wait so long?

ibuildthecloud replied:
I can change it to one second.  The 10 seconds was just copied from Container.Kill()

LK4D4 commented on daemon/delete.go:
```diff
@@ -55,7 +57,19 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
+						if _, err := container.WaitStop(10 * time.Second); err != nil {
+							return job.Errorf("Failed to wait for killed container to stop, cannot remove - %v", err)
+						}
+						if container.IsRunning() {
```
I think this check is not necessary after `WaitStop`.

ibuildthecloud replied:
I'll remove this

LK4D4 commented on daemon/delete.go:
```diff
@@ -55,7 +57,19 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
+						if container.IsRunning() {
+							return job.Errorf("Failed to kill container, cannot remove - still running")
+						}
+					} else {
```
I'm not sure, but I am personally prefer to use `early returns` without `else` clause.
Like
```
if err != execdriver.ErrProcessNotFound {
    return job.Errorf("Could not kill running container, cannot remove - %v", err)
}
```

ibuildthecloud replied:
Sure, I'll change.

ibuildthecloud commented:
@LK4D4 me too actually.  After I wrote it I realized it was a bad idea, but thought I'd still ask.  I'll update the PR and remove the masking

ibuildthecloud committed:
Wait for processes to be reaped if you are doing `rm -f`

If you do `rm -f` at just the right time when a container PID 1
has died but has not yet been reaped by Docker/Linux then kill
will return a ESRCH.  In this situation we wait a bit for
Docker/Linux to cleanup so we can delete the container.

As this is a race condition it is very difficult to reproduce but
can be done by writing a program that uses the Docker API and
does the equilavent of `docker rm -f $(docker run -d busybox echo hi)`.
Running that on the command line will never fail, but a program is
fast enough that it will hit situation a good amount of times.

Signed-off-by: Darren Shepherd <darren@rancher.com>

ibuildthecloud commented:
Updated code based on @LK4D4 feedback.  The change now is quite small (which is much better).

crosbymichael commented:
WaitStop does not actually wait4 the pid though, how does this get reaped?

LK4D4 commented:
ping @ibuildthecloud

ibuildthecloud commented:
@crosbymichael I believe the issue is that the container PID process has died but wait() has yet to return.  So it is a very small window.  According to the kill(2) man page

```
ESRCH  The pid or process group does not exist.  Note that an existing process might be a zombie, a process which already committed termination, but has not yet been wait(2)ed for.
```
So I'm calling WaitStop not to issue a wait4() but to give Docker a little bit of time to do it's job.

ibuildthecloud commented:
@jfrazelle Can I please get 1.6.0 tag on this?

jfrazelle commented:
ping @crosbymichael

LK4D4 commented:
@ibuildthecloud I'm okay with this if it fixes your problem.
LGTM

LK4D4 commented on daemon/delete.go:
```diff
@@ -55,7 +57,15 @@ func (daemon *Daemon) ContainerRm(job *engine.Job) engine.Status {
 			if forceRemove {
 				if err := container.Kill(); err != nil {
-					return job.Errorf("Could not kill running container, cannot remove - %v", err)
+					if err != syscall.ESRCH {
```
There is same check in `container.Kill`, it can't return `syscall.ESRCH`. Maybe something changed since your PR.

ibuildthecloud replied:
I will retest. That would be great if this condition doesn't happen anymore.

LK4D4 commented:
Also there is severe bug in master about this - some race between `Kill` and `container.Running`.

LK4D4 commented:
It is fixed in my branch with removing `execdriver` https://github.com/LK4D4/docker/tree/replace_execdriver , but I'm not sure when we'll merge it.

ibuildthecloud commented:
@LK4D4 I tested on master and this fix does not fix the issue anymore.  The behavior is slightly different now.  If you delete at just the right time you get

```
ERRO[0074] Handler for DELETE /containers/{name:.*} returned error: Could not kill running container, cannot remove - [2] Container does not exist: container destroyed 
ERRO[0074] HTTP Error: statusCode=500 Could not kill running container, cannot remove - [2] Container does not exist: container destroyed 
```

Basically it's trying to send a "kill -9" to a process that just exited so the following code fails https://github.com/docker/docker/blob/master/daemon/execdriver/native/driver.go#L221 because you are trying to send a signal to a container with status Destroyed.

I don't know how to address this.  Previously I was just looking for ESRCH but it seems ESRCH will be effectively masked by https://github.com/docker/docker/blob/a5269223a78d963774b12359a1e505b10fa9ac81/vendor/src/github.com/docker/libcontainer/container_linux.go#L271

So my issue still exists that it's possible to do a `rm -f` at just the right time and it will fail.  Is there anyway for libcontainer's Kill() to return ESRCH and not convert that to a generic error?

ibuildthecloud commented:
Maybe an alternate approach would be to not check active.State() during kill at https://github.com/docker/docker/blob/a5269223a78d963774b12359a1e505b10fa9ac81/daemon/execdriver/native/driver.go#L221.  Then a Kill() call will alway just return proper syscall errors that aren't masked by libcontainers errors

ibuildthecloud commented:
Oh wait, i see you call state to get the PID.  Hmm.... I'm stumped

LK4D4 commented:
@ibuildthecloud Yup, master is pretty bad now. Could you try my branch from https://github.com/LK4D4/docker/tree/replace_execdriver . There is no more execdriver there and it has more atomic operations on containers.

ibuildthecloud commented:
Got the following on your branch

```
DEBU[0067] Sending 9 to 7cea99f0c819b7f6c8bfb8c8c4f5b82b361baef2e878f9079cc5268c81d0658c 
Could not kill running container, cannot remove - os: process already finished
INFO[0067] -job rm(7cea99f0c819b7f6c8bfb8c8c4f5b82b361baef2e878f9079cc5268c81d0658c) = ERR (1) 
ERRO[0067] Handler for DELETE /containers/{name:.*} returned error: Could not kill running container, cannot remove - os: process already finished 
ERRO[0067] HTTP Error: statusCode=500 Could not kill running container, cannot remove - os: process already finished 
```

I'll dig into what's happening

ibuildthecloud commented:
I think that is from https://github.com/golang/go/blob/ecd630de6dd2e5285f18c650207b5c84649402db/src/os/exec_unix.go#L55

So golang sees ESRCH and then masks it with their own private error.

ibuildthecloud commented:
@LK4D4 after using your branch plus https://github.com/docker/libcontainer/pull/457, it is working for me.

LK4D4 commented:
@ibuildthecloud Can you confirm that in this form this PR won't affecting current behavior and can be closed?

aluzzardi commented:
@ibuildthecloud Friendly ping on @LK4D4 comment

ibuildthecloud commented:
This PR is no long valid after fixes in libcontainer.
